Adding more functionality to contrib/ansible

jeffbean opened the pull request:
Added flannel role to config flannel networking
Added docker role for more control on the daemon
Added kubernetes-addon role to mimic more of the saltbase

Working to get all the addons working, but watching #8107 as it might change how they are installed/upgraded

Verified for CentOS 7 installs.

xiang90 committed:
pkg/apiserver: minor cleanup for errToAPIStatus()

Commit:
implementation of proxy port allocation

cjcullen committed:
Add an advertise-address flag. This allows the address that the apiserver binds
to (possibly 0.0.0.0) to be different than the address on which members of the cluster
can reach the apiserver (possibly not a local interface).

smarterclayton committed:
Gracefully delete pods from the Kubelet

This commit wires together the graceful delete option for pods
on the Kubelet.  When a pod is deleted on the API server, a
grace period is calculated that is based on the
Pod.Spec.TerminationGracePeriodInSeconds, the user's provided grace
period, or a default.  The grace period can only shrink once set.
The value provided by the user (or the default) is set onto metadata
as DeletionGracePeriod.

When the Kubelet sees a pod with DeletionTimestamp set, it uses the
value of ObjectMeta.GracePeriodSeconds as the grace period
sent to Docker.  When updating status, if the pod has DeletionTimestamp
set and all containers are terminated, the Kubelet will update the
status one last time and then invoke Delete(pod, grace: 0) to
clean up the pod immediately.

smarterclayton committed:
Handle deleted pods in replication and endpoint controllers

Pods that are slated for deletion should be excluded from
replication and endpoints immediately.

smarterclayton committed:
Enable network-tester to test graceful deletion

cjcullen committed:
Clarify description/usage of --advertise-address, Master.PublicAddress

smarterclayton committed:
Show terminating pods in get/describe

vishh committed:
Avoid making connections to the apiserver insecure by default in kube2sky.

smarterclayton committed:
Verify in hack/test-cmd

smarterclayton committed:
Expire is a new type of watch action

piosz committed:
Re-enabled load e2e test for Jenkins scalability suite

Also increased next pod startup waiting time from 100s to 200s
to eliminate flakiness by incresing failCount param.

smarterclayton committed:
Delete resources immediately from e2e tests

smarterclayton committed:
Namespace controller must wait for terminating resources

erictune committed:
Updated node.md

Describe kubelet self-registration.
Remove references to deprecated node controller flags.
Describe flags that control kubelet registration.

cjcullen committed:
Use Node IP Address instead of Node.Name in minion.ResourceLocation.
Refactor GetNodeHostIP into pkg/util/node (instead of pkg/util to break import cycle).

Include internalIP in gce NodeAddresses. Remove NodeLegacyHostIP

jszczepkowski committed:
Added e2e test case for network partition.

Added e2e test case which verifies if a node can return to cluster after longer network partition. Valid for gce. Finally fixes to #7051.

Commit:
Add support for pluggable Docker exec handlers

Add support for pluggable Docker exec handlers. The default handler is
now Docker's native exec API call. The previous default, nsenter, can be
selected by passing --docker-exec-handler=nsenter when starting the
kubelet.

smarterclayton committed:
Ensure TTL is not cleared in possible edge cases

brendandburns committed:
Add a test for the service load balancer.

yujuhong committed:
Adapt most CommandRunner tests to use FakeRuntime

smarterclayton committed:
Add a new contiguous allocator strategy option

Also fix an error in port allocator

cjcullen committed:
Add an explicit KUBERNETES_MASTER variable to the kube-env.

fgrzadkowski committed:
Reduce number of list requests in watch-based cache

vishh committed:
New release of kube2sky v1.8

derekwaynecarr committed:
Document admission control plug-ins

thockin committed:
Make services.sh e2e more reliable(?) and debuggable

rjnagal committed:
Fix guestbook example based on hackathon feedback.

. Changed all kubectl invocations to be kubectl instead of cluster/kubectl.sh in some places.
. Remove mention of journalctl - no one knew what it was.
. Add create external load balancer to make it work on GCE by default. Mentioned removing it if running locally.
. Change port mentioned in doc from 8000 to 80.
. Example cleanup should not suggest tearing down the whole cluster.

erictune committed:
Explain that file-based pods cannot use secrets.

vmarmol committed:
Introduce SystemContainer to Kubelet ContainerManager.

This generalizes the handling of containers in the
ContainerManager.

Also introduces the ability to determine how much
resources are reserved for those system containers.

WIZARD-CXY committed:
bump the k8s version to 0.18.0

kazegusuri committed:
validate all examples

runningwild committed:
Change isVolumeConflict so read-only gce persistent disks don't conflict with each other

fgrzadkowski committed:
Use watch-based cache in density e2e tests.

jsafrane committed:
Fix persistent volumes.

Check Spec.PersistentVolumeSource in NFS, RBD, Gluster and iSCSI volume
plugins.

smarterclayton committed:
Apply fix for #8642 to GracefulDeletion

markturansky committed:
fixed test

Commit:
Don't assume that CA_CERT_BASE64 will be set.

thockin committed:
Insert 'svc' into the DNS search paths

Fixes #8569.

This requires the DNS server to be running kube2sky v1.6 or higher (part of
release 0.18).  Users with older kube2sky MUST NOT update to this kubelet until
they upgrade DNS.  Versions of kube2sky >= 1.6 support both old and new style
names.  Old style names are deprectaed and will be removed around the time of
kubernetes v1.0 release.

yujuhong committed:
Move more docker-specific tests to dockertools/manager_test.go

yujuhong committed:
Switch more tests from newTestKubelet to newTestKubeletWithFakeRuntime

thockin committed:
Document the supported DNS schema

Commit:
add v1 swagger-spec; update existing swagger-spec

yujuhong committed:
Make PodList span multiple lines for readability

derekwaynecarr committed:
Align systemd environ file with optimal admission controllers

jeffbean committed:
Adding more roles to ansible contrib

Added flannel role to config flannel networking
Added docker role for more control on the daemon
Added kubernetes-addon role to mimic more of the saltbase

jeffbean committed:
Prepare for 0.18.0 apiserver change.

liggitt committed:
fix omitempty on service account in v1beta3

Commit:
Version bump to 0.18.0

Kargakis committed:
expose: Better error formatting and generic flag message

Re error formatting:
$ osc expose svc frontend --port=5432 --generator='service/v2' --name=test

Before:
error: generator %!!(MISSING)q(<nil>) not found.
see 'osc expose -h' for help.

Now:
error: generator "services/v2" not found.
see 'openshift cli expose -h' for help.

fgrzadkowski committed:
Retry downloading binary tars in case of unpack failure

piosz committed:
Creating RCs in batches not in parallel in Load e2e test

zmerlynn committed:
gke provider: Fix "gcloud .* describe" checks in util.sh

This basically prevents you from re-running e2es cross-project on the
GKE provider. (The current Jenkins jobs that have been moved to
different projects are working kind of accidentally.)

piosz committed:
Disabled load e2e test for scalability Jenkins job

eparis commented:
Can you check out #9041 ? i'll probably ask you to rebase on that. I'm now liking the idea of using tags instead of seperate top level files, but we can get to that...

eparis commented on contrib/ansible/group_vars/all.yml:
```diff
@@ -10,8 +10,41 @@ ansible_ssh_user: root
 # addresses do not need to be routable and must just be an unused block of space.
 kube_service_addresses: 10.254.0.0/16
+
+flannel_network: 192.168.0.0
```
Document?

eparis commented on contrib/ansible/roles/common/tasks/main.yml:
```diff
@@ -4,6 +4,16 @@
+
+- name: Install Testing centos7 repo for new tool versions
+  copy: src=virt7-testing.repo dest=/etc/yum.repos.d/virt7-testing.repo
+  when: is_centos7
```
Lets move this repo copy to its own task.  In case other distros want to do stuff I don't want to keep cluttering this file...

eparis commented on contrib/ansible/roles/docker/tasks/main.yml:
```diff
@@ -0,0 +1,30 @@
+  tags: docker
+
+- name: Install http_proxy into docker drop-in
+  copy: dest=/usr/lib/systemd/system/docker.service.d/51-http-proxy.conf content="[Service]\nEnvironment="HTTP_PROXY={{ http_proxy }}""
```
Configuration does not belong in /usr/lib/

/etc only.

eparis commented on contrib/ansible/roles/docker/tasks/main.yml:
```diff
@@ -0,0 +1,30 @@
+- name: Install no-proxy into docker drop-in
+  copy: dest=/usr/lib/systemd/system/docker.service.d/49-no-proxy.conf content="[Service]\nEnvironment="NO_PROXY={{ no_proxy }}""
+  notify: 
+    - restart docker
```
To be honest, I don't understand what is happening here vs using /etc/sysconfig/docker like you do below...

eparis commented on contrib/ansible/roles/docker/handlers/main.yml:
```diff
@@ -0,0 +1,8 @@
+    - restart dockerservice
+
+- name: restart dockerservice
+  service: name=docker state=restarted
```
hmmm, would different naming make more sense?  maybe reload systemd and restart docker, and have the callers notify both?

eparis replied:
I don't love cascading notifications because they make handler ordering harder to read. In this case, there are only 2, so it's not so bad, but just being pedantic...

eparis commented on contrib/ansible/roles/flannel-config/templates/config-flannel.sh.j2:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+etcdctl mk /coreos.com/network/config '{"Network": "{{ flannel_network }}/16", "SubnetLen": 24, "Backend": {"Type": "vxlan"}}'
```
should we make the /16 and 24 configurable?  I have an alternate flannel implementation i haven't pushed.  Maybe we should pull this out and work on flannel in a different PR?

eparis commented on contrib/ansible/roles/flannel/meta/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
+dependencies:
+  - { role: docker }
```
that's cool, didn't know about "dependencies"

eparis commented on contrib/ansible/roles/kubernetes-addons/handlers/main.yml:
```diff
@@ -0,0 +1 @@
+---
```
do we need this file?

eparis commented on contrib/ansible/roles/kubernetes-addons/templates/kube-addons.sh:
```diff
@@ -0,0 +1,153 @@
+#!/bin/bash
```
I'm guessing this is cp'd from somewhere else in the tree?  does git support symlinks?  I don't want this getting out of date...

eparis replied:
Also, can you help me understand why this needs to be run on every boot?  Shouldn't it only need to be run once? Why is this a systemd service?

eparis commented on contrib/ansible/roles/kubernetes-addons/templates/skydns-rc.yaml.j2:
```diff
@@ -0,0 +1,66 @@
+apiVersion: v1beta3
```
I wish we could re-use these, but i'm getting the sale v ansible/jinja variables make that impossible?

jeffbean replied:
I want to run this by you before i do it. I made a replace filter to replace all the pillar vars with the same name var in Ansible. 

So the files will stay up to date but if variable names change we would have to update them.

```
- name: DNS |Download kube-addons.sh file from Kubernetes repo
  get_url: 
    url=https://raw.githubusercontent.com/GoogleCloudPlatform/kubernetes/master/cluster/addons/dns/skydns-svc.yaml.in
    dest=/etc/kubernetes/addons/dns/skydns-svc.yaml
    mode=0755 
    owner=root
    group=root
  tags: dns

- name: DNS | Convert pillar vars to ansible vars for skydns-svc.yaml
  replace: 
    dest=/etc/kubernetes/addons/dns/skydns-rc.yaml
    regexp="pillar\[\'(\w*)\'\]"
    replace="\1"
  tags: dns
```

Let me know if I am too crazy on this one.

eparis commented on contrib/ansible/roles/kubernetes/tasks/main.yml:
```diff
@@ -3,12 +3,12 @@
 
 - name: Install kubernetes
-  yum: pkg=kubernetes state=latest
+  yum: pkg=kubernetes state=latest enablerepo=virt7-testing
```
that's going to break on non-centos....

eparis commented on contrib/ansible/roles/kubernetes-addons/tasks/main.yml:
```diff
@@ -0,0 +1,20 @@
+  tags: addons
+
+- name: Install kube-addons service 
+  template: src=kube-addons.service dest=/usr/lib/systemd/system/kube-addons.service
```
belongs in /etc, not /usr/lib

eparis replied:
wouldn't we need a systemd daemon-reload here?

jeffbean commented:
Thanks for the quick feedback! 

This all helps a ton. 
I will try and get all this fixed before Kubernetes goes and changes everything again.

eparis commented:
No I think you should review #9041 for me   :)

eparis commented:
https://github.com/GoogleCloudPlatform/kubernetes/commit/066acab6fd4e670483456421675e994840b9af32

Had my thoughts on flannel. It still needs some work.

I'd like it better if etcdctl told you if it changed anything.  I think your curl stuff works.  So that's better than what I have.  Could probably use your curl + my organization to actually get updates to the config file to restart things on the nodes...

thockin committed:
Make a namespace before using it

fgrzadkowski committed:
Revert "Gracefully delete pods from the Kubelet"

Commit:
Prevent access to the /run and /exec endpoints on the master's kubelet
so that users can't run arbitrary code on the master.

bgrant0607 committed:
Revert "Add a test for the service load balancer."

jayunit100 committed:
Cleanup ginkgo declarative naming in E2Es

vmarmol committed:
Refactoring handling of latency recording.

vmarmol committed:
Add DockerErrors metric in the Kubelet.

Allows the tracking of errors by Docker operation.

bgrant0607 committed:
Revert "Make a namespace before using it"

bgrant0607 committed:
Revert "Make services.sh e2e more reliable(?) and debuggable"

bgrant0607 committed:
Revert "Document admission control plug-ins for Kube 1.0"

thockin committed:
Make services e2e test more debuggable

brendandburns committed:
Fix a bug in kubectl exec handling.

bgrant0607 committed:
Revert "Cleanup ginkgo declarative naming in E2Es"

jayunit100 committed:
Revert "Merge pull request #9166 from GoogleCloudPlatform/revert-8504-ginkgo-names"

This reverts commit 479e5005f5497e1f4488e9a79c96351be8280f28, reversing
changes made to 5115f0e6bc9069bcc9969aca244d324ed7e30023.

jayunit100 committed:
Fix bad namespace

bgrant0607 committed:
Revert "Make services e2e test more debuggable"

piosz committed:
Increased failCount param in RunRC function for e2e tests.

The previous value was 10 (equivalent of 50sec) current is 24 (2min).
Time in brackets mean: how long should I wait for the next pod of RC to be created.

While we are creating 3000 pods for scalability tests it tends to fail.

fgrzadkowski committed:
Decrease threshold for latency metrics in density tests

bgrant0607 committed:
Revert "Added e2e test case for network partition."

fgrzadkowski committed:
Load test: set threshold for latency metrics and add periodic lists
